collect_app/src/main/java/org/odk/collect/android/projects/ProjectDeleter.kt

```diff
@@ -14,39 +15,60 @@ class ProjectDeleter(
     private val formUpdateScheduler: FormUpdateScheduler,
     private val instanceSubmitScheduler: InstanceSubmitScheduler,
     private val instancesRepository: InstancesRepository,
-    private val projectDirPath: String
+    private val projectDirPath: String,
+    private val changeLockProvider: ChangeLockProvider
 ) {
     fun deleteCurrentProject(): DeleteProjectResult {
-        if (instancesRepository.getAllByStatus(
-                Instance.STATUS_INCOMPLETE,
-                Instance.STATUS_COMPLETE,
-                Instance.STATUS_SUBMISSION_FAILED
-            ).isNotEmpty()
-        ) {
-            return DeleteProjectResult.UnsentInstances
-        } else {
-            val currentProject = currentProjectProvider.getCurrentProject()
+        return when {
+            unsentInstancesDetected() -> DeleteProjectResult.UnsentInstances
+            runningBackgroundJobsDetected() -> DeleteProjectResult.RunningBackgroundJobs
+            else -> performProjectDeletion()
```

Comment:
Minor nitpick: this can just match `DeleteProjectResult.DeletedSuccessfully` instead. Then if you add a new subtype the `when` will fail to compile (which is a nice piece of protection).